packages/frontend/src/diagram/diagram_editor.tsx

```diff
@@ -53,18 +54,20 @@ export function DiagramDocumentEditor(props: {
 }) {
     return (
         <div class="growable-container">
-            <Toolbar>
-                <DiagramMenu liveDiagram={props.liveDiagram} />
-                <span class="filler" />
-                <TheoryHelpButton theory={props.liveDiagram?.liveModel.theory()} />
-                <MaybePermissionsButton
-                    permissions={props.liveDiagram?.liveDoc.permissions}
-                    refId={props.liveDiagram?.refId}
-                />
-            </Toolbar>
-            <Show when={props.liveDiagram}>
-                {(liveDiagram) => <DiagramPane liveDiagram={liveDiagram()} />}
-            </Show>
+            <ErrorBoundary fallback={(err) => <ErrorBoundaryDialog error={err} />}>
```

Comment:
A big question I have is where should the error boundaries go? Currently we've added them to the model and diagram editor components. Is there a reason not to just stick an error boundary at the top level of the app?

Reply:
They should go at the top level of the app, and adding more of them further down the component hierarchy can help improve UX when errors happen.

Outside of the top level, the heuristic I use for wrapping a component in an error boundary is to consider whether users still have viable actions on the page if the component fails. This typically means wrapping the root of navigation components like panels, and "app" components like the model/diagram editor. This increases the chance that the user can easily recover without reloading or closing the page. In my experience pages typically end up with somewhere between 1 and 4 error boundaries.

For now I have moved the error boundary to the top level.

Reply:
Thanks for that explanation, Jason. Makes sense.